test/jdk/java/lang/invoke/VarHandles/LazyInitializingTest.java

```diff
@@ -0,0 +1,204 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8291065
+ * @summary Checks interaction of static field VarHandle with class
+ *          initialization mechanism..
+ * @run junit LazyInitializingTest
+ * @run junit/othervm -Djava.lang.invoke.VarHandle.VAR_HANDLE_IDENTITY_ADAPT=true LazyInitializingTest
+ */
+
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.lang.constant.ConstantDescs;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.Objects;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class LazyInitializingTest {
+
+    record SampleData(Runnable callback, int initialValue) {
+        private static final Runnable FAIL_ON_CLINIT_CALLBACK = () -> {
+            throw new AssertionError("Class shouldn't be initialized");
+        };
+        static final SampleData FAIL_ON_CLINIT = new SampleData();
+
+        SampleData() {
+            this(FAIL_ON_CLINIT_CALLBACK, 0);
+        }
+    }
+    record ClassInfo(MethodHandles.Lookup definingLookup, VarHandle vh) {}
+
+    private static final MethodHandles.Lookup LOOKUP = MethodHandles.lookup();
+
+    /**
+     * Meta test to ensure the testing mechanism to check initialization is correct.
+     */
+    @Test
+    public void testMeta() throws IllegalAccessException {
+        boolean[] val = new boolean[1];
+        var v0 = createSampleClass(new SampleData(() -> val[0] = true, 0));
+        assertFalse(val[0], "callback run before class init");
+        v0.definingLookup.ensureInitialized(v0.definingLookup.lookupClass());
+        assertTrue(val[0], "callback not run at class init");
+    }
+
+    @Test
+    public void testUninitializedOperations() {
+        var ci = createSampleClass(SampleData.FAIL_ON_CLINIT);
+        var vh = ci.vh;
+        vh.describeConstable();
+        vh.isAccessModeSupported(VarHandle.AccessMode.GET_AND_BITWISE_XOR_ACQUIRE);
+        vh.withInvokeExactBehavior();
+        vh.withInvokeBehavior();
+        vh.toMethodHandle(VarHandle.AccessMode.GET_AND_BITWISE_XOR_ACQUIRE);
+        vh.hasInvokeExactBehavior();
+        vh.accessModeType(VarHandle.AccessMode.GET_AND_BITWISE_XOR_ACQUIRE);
+    }
+
+    @Test
+    public void testInitializationOnVarHandleUse() {
```

Comment:
I think we will need to expand this to also test operating on a MH to a VH, an indirect VH, and on a MH to indirect VH. We could do this separately.

Reply:
The MH on VH case is worth considering; the indirect VH cases are already covered by `-Djava.lang.invoke.VarHandle.VAR_HANDLE_IDENTITY_ADAPT=true`.

Reply:
Done; added a test case to test initialization on toMethodHandle use in latest push, passed locally.

Reply:
Ok. I just realized you could also run using the system property `java.lang.invoke.VarHandle.VAR_HANDLE_GUARDS` to disables the guards, and it will go through the MH invokers, achieving the same thing. Up to you if you want to keep the current test or use the system property.